zebrad/src/commands/start/sync.rs

```diff
@@ -0,0 +1,208 @@
+#![allow(unused_variables, dead_code)]
+use color_eyre::Report;
+use eyre::eyre;
+use futures::stream::{FuturesUnordered, StreamExt};
+use std::{collections::HashSet, iter};
+use tower::{Service, ServiceExt};
+use tracing_futures::Instrument;
+use zebra_chain::{block::BlockHeaderHash, types::BlockHeight};
+
+pub struct Syncer<ZN, ZS>
+where
+    ZN: Service<zebra_network::Request>,
+{
+    pub peer_set: ZN,
+    pub state: ZS,
+    pub tip_requests: FuturesUnordered<ZN::Future>,
+    pub block_requests: FuturesUnordered<ZN::Future>,
+    pub block_locator: Vec<BlockHeaderHash>,
+    pub downloading: HashSet<BlockHeaderHash>,
+    pub downloaded: HashSet<BlockHeaderHash>,
+    // TODO(jlusby): move to a config
+    pub fanout: u32,
+}
+
+impl<ZN, ZS> Syncer<ZN, ZS>
+where
+    ZN: Service<zebra_network::Request, Response = zebra_network::Response, Error = Error>
+        + Send
+        + Clone
+        + 'static,
+    ZN::Future: Send,
+    ZS: Service<zebra_state::Request, Response = zebra_state::Response, Error = Error>
+        + Send
+        + Clone
+        + 'static,
+    ZS::Future: Send,
+{
+    pub async fn run(&mut self) -> Result<(), Report> {
+        loop {
+            if self.tip_requests.is_empty() {
+                info!("populating prospective chain list");
+                self.populate_prospectives(vec![super::GENESIS]).await?;
+            }
+
+            info!("extending prospective chains");
+            self.extend_chains().await?;
```

Comment:
tioli: I think that `get_prospective_tips` and `extend_tips` would be better terminology than `populate_prospectives` and `extend_chains`, because I think they make it more clear that these functions are working just on the *tips* and not on the chain (e.g., they don't touch the existing chain in the state sequence, they don't care about block orderings, etc.)

Reply:
(edited to `obtain_tips` to match the description)